### DISCUSSION TITLE
Argument $abstract to Container::resolve?

### DISCUSSION THREAD

**Discussion by ptomulik** [laravel/framework#45852] [Q&A]
Hi, I'm trying to understand some internals of Laravel Container.  I've seen, the parameter **$abstract**  to  ``Container::resolve`` as well as ``Container::make`` is [declared to be ``string|callable``](https://github.com/laravel/framework/blob/bdbd1aa6145f6a88244eb8f78e08e332dcf0e3b1/src/Illuminate/Container/Container.php#L730). 

What's the use of ``callable`` there? Is it intentional? The **$abstract** [is used within ``resolve()`` as a key to access array items](https://github.com/laravel/framework/blob/bdbd1aa6145f6a88244eb8f78e08e332dcf0e3b1/src/Illuminate/Container/Container.php#L1301). I guess, this is not supposed to work (supposedly, it will result with ``Fatal error: Illegal offset type``).

So my question is: is this ``callable`` intentional, or it's there just by accident?

**Comment by henzeb** [comment]
It is for the bindings that use closures to build up concrete classes. the resolve method sees this as `buildable` and will execute that closure first before building up the concrete class.

See: https://laravel.com/docs/9.x/container#simple-bindings

**Comment by ptomulik** [comment]
Thank you for the answer. I'm still wondering how it's supposed to work. I've tried to use it with a closure and it doesn't seem to work. Note, that binding a callable doesn't mean it's passed to ``resolve``/``make``.  I made a simple experiment with the ``make`` (it just passes arguments to ``resolve``)  using psysh:

```psysh
> namespace Illuminate\Container;
> $container = new Container();
> $e = $container->make(\Exception::class);
= Exception {#2734
    #file: "/home/ptomulik/sources/laravel-framework/src/Illuminate/Container/Container.php",
    #line: 929,
  }
> $container->make(function ($app) { return new \Exception(); });

    TypeError  Illegal offset type in isset or empty.

```
The type error is (probably) thrown from [here](https://github.com/laravel/framework/blob/bdbd1aa6145f6a88244eb8f78e08e332dcf0e3b1/src/Illuminate/Container/Container.php#L1301)
